File: src/edge-agent/app/src/main/kotlin/com/fccmiddleware/edge/peer/PeerApiModels.kt
package com.fccmiddleware.edge.peer

import kotlinx.serialization.Serializable

// ── Heartbeat ───────────────────────────────────────────────────────────────

@Serializable
data class PeerHeartbeatRequest(
    val agentId: String,
    val siteCode: String,
    val currentRole: String,
    val leaderEpoch: Long,
    val leaderAgentId: String? = null,
    val configVersion: String? = null,
    val replicationLagSeconds: Double = 0.0,
    val lastSequenceApplied: Long = 0,
    val deviceClass: String,
    val appVersion: String,
    val uptimeSeconds: Long,
    val sentAtUtc: String,
)

@Serializable
data class PeerHeartbeatResponse(
    val agentId: String,
    val currentRole: String,
    val leaderEpoch: Long,
    val accepted: Boolean,
    val receivedAtUtc: String,
)

// ── Health ──────────────────────────────────────────────────────────────────

@Serializable
data class PeerHealthResponse(
    val agentId: String,
    val siteCode: String,
    val currentRole: String,
    val leaderEpoch: Long,
    val fccReachable: Boolean,
    val uptimeSeconds: Long,
    val appVersion: String,
    val highWaterMarkSeq: Long,
    val reportedAtUtc: String,
)

// ── Leadership Claim ────────────────────────────────────────────────────────

@Serializable
data class PeerLeadershipClaimRequest(
    val candidateAgentId: String,
    val proposedEpoch: Long,
    val priority: Int,
    val siteCode: String,
)

@Serializable
data class PeerLeadershipClaimResponse(
    val accepted: Boolean,
    val reason: String? = null,
    val currentEpoch: Long,
)

// ── Pre-Auth Proxy ──────────────────────────────────────────────────────────

@Serializable
data class PeerProxyPreAuthRequest(
    val pumpNumber: Int,
    val nozzleNumber: Int,
    val productCode: String,
    val requestedAmount: Long,
    val unitPrice: Long,
    val currency: String,
    val odooOrderId: String,
    val vehicleNumber: String? = null,
    val customerName: String? = null,
    val customerTaxId: String? = null,
    val customerBusinessName: String? = null,
    val attendantId: String? = null,
    val correlationId: String? = null,
)

@Serializable
data class PeerProxyPreAuthResponse(
    val success: Boolean,
    val preAuthId: String? = null,
    val fccCorrelationId: String? = null,
    val fccAuthorizationCode: String? = null,
    val failureReason: String? = null,
    val status: String,
)

// ── Pump Status Proxy ───────────────────────────────────────────────────────

@Serializable
data class PeerProxyPumpStatusResponse(
    val pumps: List<PeerPumpStatus>,
)

@Serializable
data class PeerPumpStatus(
    val pumpNumber: Int,
    val status: String,
    val currentNozzle: Int? = null,
    val currentProductCode: String? = null,
)

File: src/desktop-edge-agent/src/FccDesktopAgent.Core/Peer/PeerApiModels.cs
namespace FccDesktopAgent.Core.Peer;

// ── Heartbeat ───────────────────────────────────────────────────────────────

public sealed class PeerHeartbeatRequest
{
    public string AgentId { get; set; } = string.Empty;
    public string SiteCode { get; set; } = string.Empty;
    public string CurrentRole { get; set; } = string.Empty;
    public long LeaderEpoch { get; set; }
    public string? LeaderAgentId { get; set; }
    public string? ConfigVersion { get; set; }
    public double ReplicationLagSeconds { get; set; }
    public long LastSequenceApplied { get; set; }
    public string DeviceClass { get; set; } = string.Empty;
    public string AppVersion { get; set; } = string.Empty;
    public long UptimeSeconds { get; set; }
    public DateTimeOffset SentAtUtc { get; set; }
}

public sealed class PeerHeartbeatResponse
{
    public string AgentId { get; set; } = string.Empty;
    public string CurrentRole { get; set; } = string.Empty;
    public long LeaderEpoch { get; set; }
    public bool Accepted { get; set; }
    public DateTimeOffset ReceivedAtUtc { get; set; } = DateTimeOffset.UtcNow;
}

// ── Health ──────────────────────────────────────────────────────────────────

public sealed class PeerHealthResponse
{
    public string AgentId { get; set; } = string.Empty;
    public string SiteCode { get; set; } = string.Empty;
    public string CurrentRole { get; set; } = string.Empty;
    public long LeaderEpoch { get; set; }
    public bool FccReachable { get; set; }
    public long UptimeSeconds { get; set; }
    public string AppVersion { get; set; } = string.Empty;
    public long HighWaterMarkSeq { get; set; }
    public DateTimeOffset ReportedAtUtc { get; set; } = DateTimeOffset.UtcNow;
}

// ── Leadership Claim ────────────────────────────────────────────────────────

public sealed class PeerLeadershipClaimRequest
{
    public string CandidateAgentId { get; set; } = string.Empty;
    public long ProposedEpoch { get; set; }
    public int Priority { get; set; }
    public string SiteCode { get; set; } = string.Empty;
}

public sealed class PeerLeadershipClaimResponse
{
    public bool Accepted { get; set; }
    public string? Reason { get; set; }
    public long CurrentEpoch { get; set; }
}

// ── Pre-Auth Proxy ──────────────────────────────────────────────────────────

public sealed class PeerProxyPreAuthRequest
{
    public int PumpNumber { get; set; }
    public int NozzleNumber { get; set; }
    public string ProductCode { get; set; } = string.Empty;
    public long RequestedAmount { get; set; }
    public long UnitPrice { get; set; }
    public string Currency { get; set; } = string.Empty;
    public string OdooOrderId { get; set; } = string.Empty;
    public string? VehicleNumber { get; set; }
    public string? CustomerName { get; set; }
    public string? CustomerTaxId { get; set; }
    public string? CustomerBusinessName { get; set; }
    public string? AttendantId { get; set; }
    public string? CorrelationId { get; set; }
}

public sealed class PeerProxyPreAuthResponse
{
    public bool Success { get; set; }
    public string? PreAuthId { get; set; }
    public string? FccCorrelationId { get; set; }
    public string? FccAuthorizationCode { get; set; }
    public string? FailureReason { get; set; }
    public string Status { get; set; } = string.Empty;
}

// ── Pump Status Proxy ───────────────────────────────────────────────────────

public sealed class PeerProxyPumpStatusResponse
{
    public List<PeerPumpStatus> Pumps { get; set; } = [];
}

public sealed class PeerPumpStatus
{
    public int PumpNumber { get; set; }
    public string Status { get; set; } = string.Empty;
    public int? CurrentNozzle { get; set; }
    public string? CurrentProductCode { get; set; }
}
